File: apps/agentic-healthcare/langgraph/parsers.py
"""3-tier blood-test marker extraction — faithful port from parsers.ts.

Tier 1: HTML table parsing (standard lab panels)
Tier 2: Title + FormKeysValues pairs (Romanian/European lab format)
Tier 3: Free-text fallback (tab / multi-space separated)
"""

from __future__ import annotations

import re
from dataclasses import asdict, dataclass


@dataclass
class Marker:
    name: str
    value: str
    unit: str
    reference_range: str
    flag: str

    def to_dict(self) -> dict[str, str]:
        return asdict(self)


# ── flag computation ─────────────────────────────────────────────────


def compute_flag(value: str, reference_range: str) -> str:
    try:
        num = float(value.replace(",", "."))
    except (ValueError, AttributeError):
        return "normal"

    # "undetectable / negative" references
    if re.search(r"nedetectabil|undetectable|negativ|negative", reference_range, re.I):
        return "high" if num > 0 else "normal"

    # Less-than threshold
    lt = re.match(r"^[<＜≤]\s*([\d.,]+)", reference_range)
    if lt:
        return "high" if num >= float(lt.group(1).replace(",", ".")) else "normal"

    # Greater-than threshold
    gt = re.match(r"^[>＞≥]\s*([\d.,]+)", reference_range)
    if gt:
        return "low" if num <= float(gt.group(1).replace(",", ".")) else "normal"

    # Range: lo – hi
    rng = re.search(r"([\d.,]+)\s*[-–]\s*([\d.,]+)", reference_range)
    if rng:
        lo = float(rng.group(1).replace(",", "."))
        hi = float(rng.group(2).replace(",", "."))
        if num < lo:
            return "low"
        if num > hi:
            return "high"

    return "normal"


# ── helpers ──────────────────────────────────────────────────────────


def _strip_html(html: str) -> str:
    text = re.sub(r"<br\s*/?>", "\n", html, flags=re.I)
    text = re.sub(r"<[^>]+>", "", text)
    text = text.replace("&amp;", "&").replace("&lt;", "<").replace("&gt;", ">").replace("&nbsp;", " ")
    return text.strip()


def _dedupe(markers: list[Marker]) -> list[Marker]:
    seen: set[str] = set()
    out: list[Marker] = []
    for m in markers:
        if m.name not in seen:
            seen.add(m.name)
            out.append(m)
    return out


# ── Tier 1: HTML tables ──────────────────────────────────────────────


def parse_html_table(html: str) -> list[Marker]:
    markers: list[Marker] = []
    for row_match in re.finditer(r"<tr[^>]*>([\s\S]*?)</tr>", html, re.I):
        cells = [_strip_html(c) for c in re.findall(r"<t[dh][^>]*>([\s\S]*?)</t[dh]>", row_match.group(1), re.I)]
        if len(cells) < 2:
            continue
        name = cells[0]
        value = cells[1]
        unit = cells[2] if len(cells) > 2 else ""
        reference_range = cells[3] if len(cells) > 3 else ""

        if not name or not re.search(r"\d", value) or re.match(r"^\d", name):
            continue

        markers.append(Marker(
            name=name.strip(),
            value=value.strip(),
            unit=unit.strip(),
            reference_range=reference_range.strip(),
            flag=compute_flag(value, reference_range),
        ))
    return markers


# ── Tier 2: Title + FormKeysValues ───────────────────────────────────

_SKIP_PATTERN = re.compile(
    r"^(RECOLTAT|LUCRAT|GENERAT|CNP|ADRESA|TRIMIS|ANTECEDENT|Data|Pagina)", re.I
)


def parse_form_key_values(elements: list[dict]) -> list[Marker]:
    markers: list[Marker] = []
    i = 0
    while i < len(elements) - 1:
        el = elements[i]
        nxt = elements[i + 1]

        if el.get("type") in ("Title", "NarrativeText") and nxt.get("type") == "FormKeysValues":
            name = (el.get("text") or "").strip()
            value_text = (nxt.get("text") or "").strip()

            if not name or not value_text or _SKIP_PATTERN.search(value_text):
                i += 1
                continue

            vm = re.search(r"([\d.,]+)\s*([\w/µ%µgLdlUIuimlog]+)", value_text)
            if not vm:
                i += 1
                continue

            value = vm.group(1)
            unit = vm.group(2)

            refs = list(re.finditer(r"\(([^)]+)\)", value_text))
            reference_range = refs[-1].group(1) if refs else ""

            markers.append(Marker(
                name=name,
                value=value,
                unit=unit,
                reference_range=reference_range,
                flag=compute_flag(value, reference_range),
            ))
            i += 2  # skip the FormKeysValues element
        else:
            i += 1

    return markers


# ── Tier 3: Free-text fallback ───────────────────────────────────────

_TEXT_PATTERN = re.compile(
    r"^([A-Za-z\u00C0-\u00FF][A-Za-z\u00C0-\u00FF0-9 \-/().]+?)"
    r"\s{2,}([\d.,]+)"
    r"\s+([\w/µ%µgLdlUIuIU]+)"
    r"\s+([\d.,]+\s*[-–]\s*[\d.,]+|[<>≤≥＜＞]\s*[\d.,]+)",
    re.MULTILINE,
)


def parse_text_markers(text: str) -> list[Marker]:
    markers: list[Marker] = []
    for m in _TEXT_PATTERN.finditer(text):
        name, value, unit, reference_range = m.group(1), m.group(2), m.group(3), m.group(4)
        markers.append(Marker(
            name=name.strip(),
            value=value.strip(),
            unit=unit.strip(),
            reference_range=reference_range.strip(),
            flag=compute_flag(value, reference_range),
        ))
    return markers


# ── Orchestrator ─────────────────────────────────────────────────────


def parse_markers(elements: list[dict]) -> list[Marker]:
    """Extract blood markers using the 3-tier strategy."""
    # 1. HTML tables
    table_markers: list[Marker] = []
    for el in elements:
        if el.get("type") == "Table" and el.get("metadata", {}).get("text_as_html"):
            table_markers.extend(parse_html_table(el["metadata"]["text_as_html"]))
    if table_markers:
        return _dedupe(table_markers)

    # 2. Title + FormKeysValues
    fkv = parse_form_key_values(elements)
    if fkv:
        return _dedupe(fkv)

    # 3. Text fallback
    text = "\n".join(el.get("text", "") for el in elements)
    return _dedupe(parse_text_markers(text))

File: apps/agentic-healthcare/app/(app)/blood-tests/parsers.ts
export type Marker = {
  name: string;
  value: string;
  unit: string;
  reference_range: string;
  flag: string;
};

function stripHtml(html: string): string {
  return html
    .replace(/<br\s*\/?>/gi, "\n")
    .replace(/<[^>]+>/g, "")
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&nbsp;/g, " ")
    .trim();
}

export function computeFlag(value: string, reference_range: string): string {
  const num = parseFloat(value.replace(",", "."));
  if (isNaN(num)) return "normal";

  // If reference is "undetectable / negative" and we have a value → high
  if (/nedetectabil|undetectable|negativ|negative/i.test(reference_range)) {
    return num > 0 ? "high" : "normal";
  }

  const ltMatch = reference_range.match(/^[<＜≤]\s*([\d.,]+)/);
  const gtMatch = reference_range.match(/^[>＞≥]\s*([\d.,]+)/);
  if (ltMatch) return num >= parseFloat(ltMatch[1].replace(",", ".")) ? "high" : "normal";
  if (gtMatch) return num <= parseFloat(gtMatch[1].replace(",", ".")) ? "low" : "normal";

  const rangeMatch = reference_range.match(/([\d.,]+)\s*[-–]\s*([\d.,]+)/);
  if (rangeMatch) {
    const low = parseFloat(rangeMatch[1].replace(",", "."));
    const high = parseFloat(rangeMatch[2].replace(",", "."));
    if (!isNaN(low) && !isNaN(high)) {
      if (num < low) return "low";
      if (num > high) return "high";
    }
  }
  return "normal";
}

/** Parse HTML tables (e.g. standard lab panels) */
function parseHtmlTable(html: string): Marker[] {
  const markers: Marker[] = [];
  const rowRegex = /<tr[^>]*>([\s\S]*?)<\/tr>/gi;
  let rowMatch;
  while ((rowMatch = rowRegex.exec(html)) !== null) {
    const cellRegex = /<t[dh][^>]*>([\s\S]*?)<\/t[dh]>/gi;
    const cells: string[] = [];
    let cellMatch;
    while ((cellMatch = cellRegex.exec(rowMatch[1])) !== null) {
      cells.push(stripHtml(cellMatch[1]));
    }
    if (cells.length < 2) continue;
    const [name, value, unit = "", reference_range = ""] = cells;
    if (!name || !/\d/.test(value) || /^\d/.test(name)) continue;
    markers.push({
      name: name.trim(),
      value: value.trim(),
      unit: unit.trim(),
      reference_range: reference_range.trim(),
      flag: computeFlag(value, reference_range),
    });
  }
  return markers;
}

/** Parse Title + FormKeysValues pairs (Romanian/European lab format) */
export function parseFormKeyValues(
  elements: Array<{ [k: string]: any }>
): Marker[] {
  const markers: Marker[] = [];

  const SKIP_PATTERN =
    /^(RECOLTAT|LUCRAT|GENERAT|CNP|ADRESA|TRIMIS|ANTECEDENT|Data|Pagina)/i;

  for (let i = 0; i < elements.length - 1; i++) {
    const el = elements[i];
    const next = elements[i + 1];

    if (
      (el.type === "Title" || el.type === "NarrativeText") &&
      next?.type === "FormKeysValues"
    ) {
      const name = el.text?.trim() ?? "";
      const valueText = next.text?.trim() ?? "";

      if (!name || !valueText || SKIP_PATTERN.test(valueText)) continue;

      // Extract first numeric token + its unit
      const valueMatch = valueText.match(/([\d.,]+)\s*([\w/µ%µgLdlUIuimlog]+)/);
      if (!valueMatch) continue;

      const value = valueMatch[1];
      const unit = valueMatch[2];

      // Last parenthetical is typically the reference
      const refs = [...valueText.matchAll(/\(([^)]+)\)/g)];
      const reference_range = refs.length > 0 ? refs[refs.length - 1][1] : "";

      markers.push({
        name,
        value,
        unit,
        reference_range,
        flag: computeFlag(value, reference_range),
      });
      i++; // skip the FormKeysValues element
    }
  }

  return markers;
}

/** Flexible line-based text parser (tab / multi-space separated) */
export function parseTextMarkers(text: string): Marker[] {
  const markers: Marker[] = [];
  const pattern =
    /^([A-Za-zÀ-ÿ][A-Za-zÀ-ÿ0-9 \-/().]+?)\s{2,}([\d.,]+)\s+([\w/µ%µgLdlUIuIU]+)\s+([\d.,]+\s*[-–]\s*[\d.,]+|[<>≤≥＜＞]\s*[\d.,]+)/gm;
  let match;
  while ((match = pattern.exec(text)) !== null) {
    const [, name, value, unit, reference_range] = match;
    markers.push({
      name: name.trim(),
      value: value.trim(),
      unit: unit.trim(),
      reference_range: reference_range.trim(),
      flag: computeFlag(value, reference_range),
    });
  }
  return markers;
}

export function parseMarkers(
  elements: Array<{ [k: string]: any }>
): Marker[] {
  // 1. HTML tables
  const tableMarkers: Marker[] = [];
  for (const el of elements) {
    if (el.type === "Table" && el.metadata?.text_as_html) {
      tableMarkers.push(...parseHtmlTable(el.metadata.text_as_html));
    }
  }
  if (tableMarkers.length > 0) return dedupe(tableMarkers);

  // 2. Title + FormKeysValues pairs (European lab format)
  const fkvMarkers = parseFormKeyValues(elements);
  if (fkvMarkers.length > 0) return dedupe(fkvMarkers);

  // 3. Text fallback
  const text = elements.map((el) => el.text ?? "").join("\n");
  return dedupe(parseTextMarkers(text));
}

function dedupe(markers: Marker[]): Marker[] {
  const seen = new Set<string>();
  return markers.filter((m) => {
    if (seen.has(m.name)) return false;
    seen.add(m.name);
    return true;
  });
}
